src/Analyzers/Core/Analyzers/RemoveRedundantEquality/AbstractRemoveRedundantEqualityDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.RemoveRedundantEquality
+{
+    internal abstract class AbstractRemoveRedundantEqualityDiagnosticAnalyzer
+        : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private readonly ISyntaxFacts _syntaxFacts;
+
+        protected AbstractRemoveRedundantEqualityDiagnosticAnalyzer(ISyntaxFacts syntaxFacts)
+            : base(IDEDiagnosticIds.RemoveRedundantEqualityDiagnosticId,
+                   option: null,
+                   new LocalizableResourceString(nameof(AnalyzersResources.Remove_redundant_equality), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)))
+        {
+            _syntaxFacts = syntaxFacts;
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterOperationAction(AnalyzeBinaryOperator, OperationKind.BinaryOperator);
+
+        private void AnalyzeBinaryOperator(OperationAnalysisContext context)
+        {
+            var operation = (IBinaryOperation)context.Operation;
+            if (operation.OperatorMethod is not null)
+            {
+                // We shouldn't report diagnostic on overloaded operator as the behavior can change.
+                return;
+            }
+
+            if (operation.OperatorKind is not (BinaryOperatorKind.Equals or BinaryOperatorKind.NotEquals))
+            {
+                return;
+            }
+
+            var rightOperand = operation.RightOperand;
+            var leftOperand = operation.LeftOperand;
+            if (rightOperand.Type.SpecialType is not SpecialType.System_Boolean ||
+                leftOperand.Type.SpecialType is not SpecialType.System_Boolean)
+            {
+                return;
+            }
+
+            var isOperatorEquals = operation.OperatorKind == BinaryOperatorKind.Equals;
+            _syntaxFacts.GetPartsOfBinaryExpression(operation.Syntax, out _, out var operatorToken, out _);
```

Comment:
this may crashif operation.Syntax is not a BinaryExpression.  We need to validate that first.

Reply:
@CyrusNajmabadi An easier alternative is to have a TryGetPartsOfBinaryExpression variant of this method.

Are you okay if I defined this method in ISyntaxFacts and implemented in CSharpSyntaxFacts and VisualBasicSyntaxFacts?

Reply:
i would not do it this way.  instead, just check if you have a BinaryExpression node.  You can use ISyntaxFacts.IsBinaryExpression for this.